segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/SegmentAggregator.java

```diff
@@ -1165,17 +1165,27 @@ private void mergeCompleted(SegmentProperties segmentProperties, UpdateableSegme
         if (this.handle.get() == null) {
             // No handle so, the segment must not exist yet. Attempt to create it, then run what we wanted to.
             assert this.metadata.getStorageLength() == 0 : "no handle yet but metadata indicates Storage Segment not empty";
-            long rolloverSize = this.metadata.getAttributes().getOrDefault(Attributes.ROLLOVER_SIZE, -1L);
-            SegmentRollingPolicy rollingPolicy = rolloverSize < 0 ? SegmentRollingPolicy.NO_ROLLING : new SegmentRollingPolicy(rolloverSize);
+            long rolloverSize = this.metadata.getAttributes().getOrDefault(Attributes.ROLLOVER_SIZE, SegmentRollingPolicy.NO_ROLLING.getMaxLength());
             return Futures
-                    .exceptionallyExpecting(
-                            this.storage.create(this.metadata.getName(), rollingPolicy, timeout),
+                    .exceptionallyComposeExpecting(
+                            this.storage.create(this.metadata.getName(), new SegmentRollingPolicy(rolloverSize), timeout),
                             ex -> ex instanceof StreamSegmentExistsException,
-                            null)
+                            () -> {
+                                // This happens if we have more than one concurrent instances of the owning SegmentContainer
+                                // running at the same time. Both SegmentAggregator instances were initialized when the Segment
+                                // did not exist, and both knew about an append that would eventually make it to Storage. One
+                                // of them managed to create the Segment (and write something to it), but the other still assumed
+                                // the Segment did not exist - so we end up in here. We need to get a handle of the segment
+                                // and continue with whatever we were doing. If there is a mismatch (length, sealed, etc.),
+                                // then the normal reconciliation algorithm will kick in once it is discovered.
+                                log.info("{}: Segment did not exist in Storage when initialized() was called, but does now.", this.traceObjectId);
+                                return this.storage.openWrite(this.metadata.getName());
```

Comment:
We are relying on the contract of `openWrite` to avoid having concurrent writes to the segment in tier 2.

Reply:
Yes. I updated the comment to reflect that.
All `Storage` operations that modify something can throw that, not just `openWrite`.

Reply:
FileSystemStorage.openWrite only checks for existence. 
See. https://github.com/pravega/pravega/blob/master/bindings/src/main/java/io/pravega/storage/filesystem/FileSystemStorage.java#L214

Reply:
What are you trying to imply @sachin-j-joshi ?

Reply:
The FileSystemStorage does not implement fencing as of today. It can not detect not being a primary storage. Therefore we can not rely on this being detected in SyncStorage.